compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -1521,4 +1868,264 @@ mod control_dependence {
 
         assert_normalized_ssa_equals(ssa, expected);
     }
+
+    #[test]
+    fn simplify_constraint() {
+        // This test shows the simplification of the constraint constrain v17 == u1 1 which is converted into constrain u1 0 == u1 1 in b0
+        let src = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32, v2: u32):
+            v4 = allocate -> &mut u32
+            store v0 at v4
+            jmp b1(u32 0)
+          b1(v3: u32):
+            v7 = lt v3, u32 5
+            jmpif v7 then: b2, else: b3
+          b2():
+            v12 = lt v3, u32 8
+            jmpif v12 then: b4, else: b5
+          b3():
+            v8 = load v4 -> u32
+            v9 = lt v1, v8
+            constrain v9 == u1 1
+            return
+          b4():
+            v13 = load v4 -> u32
+            v15 = add v13, u32 1
+            store v15 at v4
+            jmp b5()
+          b5():
+            v17 = lt v3, u32 4
+            constrain v17 == u1 1
+            v18 = unchecked_add v3, u32 1
+            jmp b1(v18)
+        }
+        ";
+
+        let ssa = Ssa::from_str(src).unwrap();
+
+        let ssa = ssa.loop_invariant_code_motion();
+        // We expect the constrain to remain inside of `loop_body`
+        // as the loop is never going to be executed.
+        // If the constrain were to be hoisted out it could potentially
+        // cause the program to fail when it is not meant to fail.
+        let expected = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32, v2: u32):
+              v4 = allocate -> &mut u32
+              store v0 at v4
+              constrain u1 0 == u1 1
```

Comment:
The comment above `expected` looks to conflict with the test here. The constrain is hoisted out while the comment says it should not be? I see we simplify out `v12 = lt v3, u32 8` so we do know that the loop will execute and we can safely hoist the constrain. Let's update the comment

Reply:
Yes sorry, it's a bad copy paste

Reply:
Done